staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -655,6 +676,63 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 	return s.versioner.UpdateList(listObj, uint64(returnedRV), "", nil)
 }
 
+func splitKVsIntoChunks(concurrencyLevel int, kvs []*mvccpb.KeyValue) [][]*mvccpb.KeyValue {
+	if concurrencyLevel == 0 {
+		return [][]*mvccpb.KeyValue{kvs}
+	}
+
+	var (
+		chunks      [][]*mvccpb.KeyValue
+		numOfChunks = concurrencyLevel
+	)
+
+	if kvs == nil || len(kvs) == 0 {
+		return nil
+	}
+
+	if concurrencyLevel > len(kvs) {
+		numOfChunks = len(kvs)
+	}
+
+	chunkSize := len(kvs) / numOfChunks
+	extra := len(kvs) % numOfChunks
+
+	start := 0
+	end := chunkSize
+	for i := 0; i < numOfChunks; i++ {
+		chunks = append(chunks, kvs[start:end])
+		start = end
+		end = end + chunkSize
+	}
+
+	// For simplicity,  appending the extra (all remaining) kv to the last bucket.
+	chunks[numOfChunks-1] = append(chunks[numOfChunks-1], kvs[len(kvs)-extra:]...)
+
+	return chunks
+}
+
+func (s *store) transformChunk(ctx context.Context, kvPairs []*mvccpb.KeyValue) <-chan *transformedKVPair {
+	outChan := make(chan *transformedKVPair, len(kvPairs))
```

Comment:
Initially, I implemented this with a single channel, but based on this [comment](https://github.com/kubernetes/kubernetes/pull/88633#discussion_r385397821), I assumed that we have to preserve the order of the kvPairs in the response. Is this correct?

In other words, the current implementation assumes that preserving the order is a requirement. Therefore, several channels are used - each processing a segment of the response slice.
These segments are later assembled, thus preserving the order of the items in the response (post transformation).

I can certainly see that we need to carefully preserve the lastKey since this is required for paging. However, I am not sure if the other kvPairs could be shuffled based on the order they arrived. Thus improving concurrency of the transformation.

Of course, we could also sort the transformed items once all of them were received. WDYT?

Reply:
We cannot risk @smarterclayton per https://github.com/openshift/origin/issues/12792#issue-205073237!

In all seriousness, list calls need to maintain the sort order from etcd.  The UX of list calls returning randomly sorted data is really annoying.

Reply:
Ordering of responses in natural etcd key order is part of our API guarantee